python/cudf/cudf/core/series.py

```diff
@@ -1635,6 +1636,71 @@ def applymap(self, udf, out_dtype=None):
           * math.gamma()
           * math.lgamma()
 
+        * Series with string dtypes are not supported in `applymap` method.
+
+        * Global variables need to be re-defined explicitly inside
+          the udf, as numba considers them to be compile-time constants
+          and there is no known way to obtain value of the global variable.
+
+        Examples
+        --------
+        Returning a Series of booleans using only a literal pattern.
+        >>> import cudf
+        >>> s = cudf.Series([1, 10, -10, 200, 100])
+        >>> s.applymap(lambda x: x)
+        0      1
+        1     10
+        2    -10
+        3    200
+        4    100
+        dtype: int64
+        >>> s.applymap(lambda x: x in [1, 100, 59])
```

Comment:
Does this actually work? I'm surprised this would JIT compile and run on the GPU nicely...

Reply:
Yep 💯 
It does work actually

Reply:
Can you do a quick smoketest on the performance?

cc @gmarkall what are your expectations here?

Reply:
`s.applymap(lambda x: x in [1027, 1000, 59, 980, 1027, 1000, 59, 980, 1027, 1000, 59, 980, 1027, 1000, 59, 980, 1027, 1000, 59, 980])`

On the above list with different number of records in `s`, here are the numbers:

| Records  | Execution Time |
| ------------- | ------------- |
| 1000  | 0.204644 s  |
| 10,000,000  | 0.093468 s  |
| 100,000,000  | 0.110157 s  |
| 200,000,000  | 0.112257 s  |

Reply:
Well that looks pretty good from my perspective, but would still prefer to hear from @gmarkall or if @hummingtree wants to dump the PTX from this function and take a look I'd feel much more comfortable before merging.

Reply:
The following test works:

```python
from numba import cuda
import numpy as np

@cuda.jit
def f(x):
    x[0] = x[0] in [1, 100, 59]


x0 = np.zeros(1, dtype=np.int64)
x1 = np.ones(1, dtype=np.int64)

f[1, 1](x0)
f[1, 1](x1)

print(x0)
print(x1)
```

and produces:

```
[0]
[1]
```

I think this is because there is some support for list operations in Nopython mode in Numba, and the CUDA target reuses a lot of the nopython mode pipeline - there are many other things supported in nopython mode that don't work on CUDA, but I think in this case we get lucky that all the relevant typing and lowering implementations get registered for the CUDA target.

PTX for the test is:

```assembly
visible .entry _ZN6cudapy8__main__5f$241E5ArrayIxLi1E1C7mutable7alignedE(
	.param .u64 _ZN6cudapy8__main__5f$241E5ArrayIxLi1E1C7mutable7alignedE_param_0,
	.param .u64 _ZN6cudapy8__main__5f$241E5ArrayIxLi1E1C7mutable7alignedE_param_1,
	.param .u64 _ZN6cudapy8__main__5f$241E5ArrayIxLi1E1C7mutable7alignedE_param_2,
	.param .u64 _ZN6cudapy8__main__5f$241E5ArrayIxLi1E1C7mutable7alignedE_param_3,
	.param .u64 _ZN6cudapy8__main__5f$241E5ArrayIxLi1E1C7mutable7alignedE_param_4,
	.param .u64 _ZN6cudapy8__main__5f$241E5ArrayIxLi1E1C7mutable7alignedE_param_5,
	.param .u64 _ZN6cudapy8__main__5f$241E5ArrayIxLi1E1C7mutable7alignedE_param_6
)
{
	.reg .pred 	%p<6>;
	.reg .b64 	%rd<5>;


	ld.param.u64 	%rd1, [_ZN6cudapy8__main__5f$241E5ArrayIxLi1E1C7mutable7alignedE_param_4];
	cvta.to.global.u64 	%rd2, %rd1;
	ld.global.u64 	%rd3, [%rd2];
	setp.eq.s64	%p1, %rd3, 1;
	setp.eq.s64	%p2, %rd3, 100;
	or.pred  	%p3, %p1, %p2;
	setp.eq.s64	%p4, %rd3, 59;
	or.pred  	%p5, %p3, %p4;
	selp.u64	%rd4, 1, 0, %p5;
	st.global.u64 	[%rd2], %rd4;
	ret;
}
```

Reply:
@kkraus14 The device PTX generated from `numba` from `lambda x: x in [0, 100, 59]` is
```//
// Generated by NVIDIA NVVM Compiler
//
// Compiler Build ID: CL-27506705
// Cuda compilation tools, release 10.2, V10.2.89
// Based on LLVM 3.4svn
//

.version 6.5
.target sm_70
.address_size 64

        // .globl       _ZN8__main__16$3clambda$3e$241Ed
.common .global .align 8 .u64 _ZN08NumbaEnv8__main__16$3clambda$3e$241Ed;
.common .global .align 8 .u64 _ZN08NumbaEnv5numba7targets7listobj6in_seq12$3clocals$3e21seq_contains_impl$242E8UniTupleIxLi3EEd;

.visible .func  (.param .b32 func_retval0) _ZN8__main__16$3clambda$3e$241Ed(
        .param .b64 _ZN8__main__16$3clambda$3e$241Ed_param_0,
        .param .b64 _ZN8__main__16$3clambda$3e$241Ed_param_1
)
{
        .reg .pred      %p<6>;
        .reg .b16       %rs<2>;
        .reg .b32       %r<2>;
        .reg .f64       %fd<2>;
        .reg .b64       %rd<2>;


        ld.param.u64    %rd1, [_ZN8__main__16$3clambda$3e$241Ed_param_0];
        ld.param.f64    %fd1, [_ZN8__main__16$3clambda$3e$241Ed_param_1];
        setp.eq.f64     %p1, %fd1, 0d0000000000000000;
        setp.eq.f64     %p2, %fd1, 0d4059000000000000;
        or.pred         %p3, %p1, %p2;
        setp.eq.f64     %p4, %fd1, 0d404D800000000000;
        or.pred         %p5, %p3, %p4;
        selp.u16        %rs1, 1, 0, %p5;
        st.u8   [%rd1], %rs1;
        mov.u32         %r1, 0;
        st.param.b32    [func_retval0+0], %r1;
        ret;
}



```
I believe it's working as expected.

Reply:
> `s.applymap(lambda x: x in [1027, 1000, 59, 980, 1027, 1000, 59, 980, 1027, 1000, 59, 980, 1027, 1000, 59, 980, 1027, 1000, 59, 980])`
> 
> On the above list with different number of records in `s`, here are the numbers:
> 
> Records	Execution Time
> 1000	0.204644 s
> 10,000,000	0.093468 s
> 100,000,000	0.110157 s
> 200,000,000	0.112257 s

The first time you run a certain python UDF it takes time to compile and it's cached. I believe that's probably the reason why your timing with 1000 rows is larger than the later ones.

Reply:
@gmarkall @hummingtree Perfect, thanks for the analysis, it's much appreciated.